File: add-two-numbers/add-two-numbers-2.go
// https://leetcode.com/problems/add-two-numbers/
// 4 ms	4.7 MB
package main
import "fmt"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct{
    Val int
    Next *ListNode
}

//----------------------------------- Answer Begin -----------------------------------//
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    root  := l1
    carry := 0
    sum   := 0
    var pre *ListNode

    for l1!=nil && l2!=nil {
        sum      = l1.Val + l2.Val + carry
        carry    = sum / 10
        l1.Val   = sum % 10
        pre      = l1
        l1       = l1.Next
        l2       = l2.Next
    }

    // if l2 is too long
    if l2!=nil {
        pre.Next = l2
        l1 = pre.Next
    }

    // if l1 is too long
    for l1!=nil {
        sum       = l1.Val + carry
        carry     = sum / 10
        l1.Val    = sum % 10
        pre       = l1
        l1        = l1.Next
    }


    if carry > 0 {
        pre.Next = new(ListNode)
        pre      = pre.Next
        pre.Val  = carry
    }

    return root
}
//----------------------------------- Answer END -----------------------------------//
func makeLL(list []int) *ListNode{
    var root *ListNode
    for i := len(list)-1; i>=0; i--{
        n     := new(ListNode)
        n.Val  = list[i]
        n.Next = root
        root   = n
    }
    return root
}

func printLL(l *ListNode){
    for cur := l; cur != nil; cur = cur.Next{
        fmt.Printf("%d -> ", cur.Val)
    }
    fmt.Printf("nil\n")
}

func main(){
    l1 := makeLL([]int{2, 4, 3, 9, 9, 9})
    printLL(l1)
    l2 := makeLL([]int{5, 6, 9, 9, 9})
    printLL(l2)
    l3 := addTwoNumbers(l1, l2)
    printLL(l3)
}

File: add-two-numbers/add-two-numbers-2.c
// https://leetcode.com/problems/add-two-numbers/
// 8 ms	7.5 MB
#include<stdio.h>
#include<stdlib.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode{
    int val;
    struct ListNode *next;
};

//----------------------------------- Answer Begin -----------------------------------//
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    struct ListNode *cur1, *cur2;
    struct ListNode *head, *tail;
    head = calloc(1, sizeof(*head));
    tail = head;
    int carry = 0;
    int sum;
    for(cur1 = l1, cur2 = l2; cur1 && cur2; cur1 = cur1->next, cur2 = cur2->next){
        sum       = cur1->val + cur2->val + carry;
        carry     = sum / 10;

        struct ListNode* new = malloc(sizeof(*new));
        new->val   = sum % 10;
        new->next  = NULL;
        tail->next = new;
        tail       = new;
    }

    for(; cur1; cur1 = cur1->next){
        sum       = cur1->val + carry;
        carry     = sum / 10;

        struct ListNode* new = malloc(sizeof(*new));
        new->val   = sum % 10;
        new->next  = NULL;
        tail->next = new;
        tail       = new;
    }

    for(; cur2; cur2 = cur2->next){
        sum       = cur2->val + carry;
        carry     = sum / 10;

        struct ListNode* new = malloc(sizeof(*new));
        new->val   = sum % 10;
        new->next  = NULL;
        tail->next = new;
        tail       = new;
    }

    if(carry > 0){
        struct ListNode* new = malloc(sizeof(*new));
        new->val  = carry;
        new->next = NULL;
        tail->next = new;
        tail       = new;
    }

    tail->next = NULL;
    return head->next;
}
//----------------------------------- Answer END -----------------------------------//

struct ListNode* makeLL(int* array, int size){
    struct ListNode* root = NULL;
    for(size--; size>=0; size--){
        struct ListNode* new = malloc(sizeof(*new));
        new -> val = array[size];
        new -> next = root;
        root = new;
    }
    return root;
}

void printLL(struct ListNode* root){
    struct ListNode* cur;
    for(cur = root; cur; cur=cur->next){
        printf("%d -> ", cur->val);
    }
    printf("NULL \n");
}

int main(){
    int arr1[] = {2,4,3,7};
    int arr2[] = {5,6,4,8};
    struct ListNode *l1 = makeLL(arr1, 4);
    struct ListNode *l2 = makeLL(arr2, 4);
    printLL(l1);
    printLL(l2);

    l1 = addTwoNumbers(l1, l2);
    printLL(l1);

    return 0;
}
